_posts/2016-04-05-bits-shapeless-1-hlists.md

```diff
@@ -0,0 +1,278 @@
+//        l.drop(4)
+//              ^
+```
+
+As the `HList`size is known in compile time (is encoded in the type itself) the `drop` is written in such a way that the compiler verifies that one is not dropping more elements than the known length. In this case it is, so the compilation fails.
+
+### Recursion with HLists
+
+suppose we want to compute the "size" of an HList and we define the size as the following:
+
+- The size of the HList is the sum of the sizes of all the hlist's elements
+- The size of an integer is 1.
+- The size of a String is the string's length
+
+(we could have defined size for other types of course, but we will only use ints and strings for our example).
+
+This translates pretty nicely to a type class [^poly], with corresponding instances for Int, String and the `HList.`
+
+In Scala, type classes are usually encoded as a trait with a generic Type and instances for concrete types being present in scope as implicit values or definitions.
+
+The `Sized` type class definition [^sized] is then as follows:
+
+```scala
+trait Sized[-T] {
+  def size(t: T): Int
+}
+
+object Sized {
+  // Helper to get the size for a value
+  def size[T](t: T)(implicit tSized: Sized[T]) = tSized.size(t)
+
+  // Helper to create a type class instance from a function from T to Int
+  def instance[T](f : T => Int) = new Sized[T] {
+    def size(t: T) = f(t)
+  }
+}
+```
+
+And instances for `int`and `String`are simply:
+
+```scala
+implicit val intSized: Sized[Int] = Sized.instance(_ => 1)
+// intSized: Sized[Int] = Sized$$anon$1@1cafc588
+
+implicit val stringSized: Sized[String] = Sized.instance(s => s.length)
+// stringSized: Sized[String] = Sized$$anon$1@4287eecb
+```
+
+Lets try it out:
+
+```scala
+import Sized.size
+// import Sized.size
+
+size(1)
+// res16: Int = 1
+
+size("hello")
+// res17: Int = 5
+```
+
+It works. However, we are still missing a way to process the HList and its elements
+
+As we have done with simple Lists, we will use recursion to process the `HList`, defining both a base (termination) case and the recursive invocation. However, instead of a runtime function with two cases (which in Scala maps to pattern matching) we will encode these in the type system with type class instances, one for `HNil` and another for the `::` recursive structure.
+
+For `HNil`we define a size of 0 (the empty list):
+
+
+```scala
+implicit val hnilSized = Sized.instance[HNil](_ => 0)
+```
+
+For `::` we define the size as the size of the head of the HList plus the size of the tail, which is also another `HList`. 
+
+```scala
+implicit def hconsSized[H, T <: HList]
+  (implicit hSized: Sized[H], tSized: Sized[T]) =
+  Sized.instance[H :: T] { l =>
+    hSized.size(l.head) + tSized.size(l.tail)
+}
+```
+
+Let's analyse this definition:
+
+- The implicit definition has two type parameters, one representing the type of the head of the HList, `H`, and `T`, which is the type of the tail of the list and is required to by a subtype of `HList`.
```

Comment:
It's confusing. replace with "...  one representing the type of the head of the HList, `H`, and the type of the tail of the HList, `T`, which are required to be a subtype of `HList`."